jetty-proxy/src/test/java/org/eclipse/jetty/proxy/ClientAuthProxyTest.java

```diff
@@ -0,0 +1,549 @@
+
+            @Override
+            protected Request newProxyRequest(HttpServletRequest request, String rewrittenTarget)
+            {
+                String user = retrieveUser(request);
+                // Obviously not fool proof, but for the 2 users in this test it does the job.
+                String subDomain = subDomains.computeIfAbsent(user, key -> "127.0.0." + userIds.incrementAndGet());
+                Request proxyRequest = super.newProxyRequest(request, rewrittenTarget);
+                proxyRequest.host(subDomain).port(serverConnector.getLocalPort());
+                // Tag the request.
+                proxyRequest.tag(new AliasTLSTag(user));
+                return proxyRequest;
+            }
+
+            @Override
+            protected HttpClient newHttpClient()
+            {
+                SslContextFactory.Client clientTLS = new SslContextFactory.Client()
+                {
+                    @Override
+                    protected KeyManager[] getKeyManagers(KeyStore keyStore) throws Exception
+                    {
+                        KeyManager[] keyManagers = super.getKeyManagers(keyStore);
+                        for (int i = 0; i < keyManagers.length; i++)
+                        {
+                            keyManagers[i] = new ProxyAliasX509ExtendedKeyManager(keyManagers[i]);
+                        }
+                        return keyManagers;
+                    }
+                };
+                // Disable TLS-level hostname verification for this test.
+                clientTLS.setEndpointIdentificationAlgorithm(null);
+                clientTLS.setKeyStorePath(MavenTestingUtils.getTestResourceFile("client_auth/proxy_keystore.p12").getAbsolutePath());
+                clientTLS.setKeyStorePassword("storepwd");
+                clientTLS.setKeyStoreType("PKCS12");
+                return new HttpClient(new HttpClientTransportOverHTTP(1), clientTLS);
+            }
+        });
+        startClient();
+
+        testRequestsFromRemoteClients();
+    }
+
+    @Test
+    public void testClientAuthProxyingWithSSLSessionResumptionDisabled() throws Exception
+    {
+        // To user the same HttpClient and server hostName, we need to disable
+        // SSLSession caching, which is only possible by creating SSLEngine
+        // without peer host information.
+        // This is more CPU intensive because TLS sessions can never be resumed.
+
+        startServer();
+        startProxy(new AsyncProxyServlet()
+        {
+            @Override
+            protected Request newProxyRequest(HttpServletRequest request, String rewrittenTarget)
+            {
+                String user = retrieveUser(request);
+                Request proxyRequest = super.newProxyRequest(request, rewrittenTarget);
+                proxyRequest.port(serverConnector.getLocalPort());
+                // Tag the request.
+                proxyRequest.tag(new AliasTLSTag(user));
```

Comment:
I still don't really like dynamic tags..... but I have reconciled myself with them by think of tags like bag tags used for routing luggage and you can now get smart bag tags that response to RF queries... so I guess this is just a Smart Tag.